pallets/pool-registry/src/benchmarking.rs

```diff
@@ -11,140 +11,201 @@
+use frame_support::traits::fungibles::Inspect;
 use frame_system::RawOrigin;
+use orml_traits::Change;
+#[cfg(feature = "runtime-benchmarks")]
+use pallet_pool_system::benchmarking::{
+	assert_input_tranches_match, assert_update_tranches_match, build_bench_input_tranches,
+	create_admin, create_investor, create_pool, get_pool, get_scheduled_update, get_tranche_id,
+	prepare_asset_registry, update_pool,
+};
+use pallet_pool_system::{
+	pool_types::PoolChanges,
+	tranches::{TrancheIndex, TrancheInput, TrancheMetadata, TrancheType, TrancheUpdate},
+};
+use sp_runtime::{
+	traits::{One, Zero},
+	Perquintill,
+};
 use sp_std::vec;
 
 use super::*;
 
+const CURRENCY: u128 = 1_000_000_000_000_000;
+const MAX_RESERVE: u128 = 10_000 * CURRENCY;
+
+const SECS_PER_HOUR: u64 = 60 * 60;
+const SECS_PER_DAY: u64 = 24 * SECS_PER_HOUR;
+const SECS_PER_YEAR: u64 = 365 * SECS_PER_DAY;
+
+const TRANCHE: TrancheIndex = 0;
+
 const POOL: u64 = 0;
 
 benchmarks! {
 	where_clause {
 	where
-		T: Config<PoolId = u64>,
+		T: Config<PoolId = u64,
+			  TrancheId = [u8; 16],
+			  Balance = u128,
+			  CurrencyId = CurrencyId> + pallet_investments::Config<InvestmentId = TrancheCurrency, Amount = u128>,
+		T: pallet_pool_system::Config<PoolId = u64,
+			  TrancheId = [u8; 16],
+			  Balance = u128,
+			  CurrencyId = CurrencyId,
+			  EpochId = PoolEpochId,
+			  Rate = <T as Config>::InterestRate,
+			  MaxTokenNameLength = <T as Config>::MaxTokenNameLength,
+			  MaxTokenSymbolLength = <T as Config>::MaxTokenSymbolLength,
+			  MaxTranches = <T as Config>::MaxTranches>,
+		<T as pallet_investments::Config>::Tokens: Inspect<T::AccountId, AssetId = CurrencyId, Balance = u128>,
+		<<T as pallet_investments::Config>::Accountant as InvestmentAccountant<T::AccountId>>::InvestmentInfo:
+			InvestmentProperties<T::AccountId, Currency = CurrencyId>,
+		<<T as frame_system::Config>::Lookup as sp_runtime::traits::StaticLookup>::Source:
+			From<<T as frame_system::Config>::AccountId>,
+		<T as pallet_pool_system::Config>::Permission: Permissions<T::AccountId, Ok = ()>,
+		<T as Config>::ModifyPool: PoolMutate<
+			<T as frame_system::Config>::AccountId,
+			<T as pallet::Config>::PoolId,
+			TrancheInput = TrancheInput<
+				<T as pallet_pool_system::Config>::Rate,
+				<T as pallet_pool_system::Config>::MaxTokenNameLength,
+				<T as pallet_pool_system::Config>::MaxTokenSymbolLength>,
+			PoolChanges = PoolChanges<
+				<T as pallet_pool_system::Config>::Rate,
+				<T as pallet_pool_system::Config>::MaxTokenNameLength,
+				<T as pallet_pool_system::Config>::MaxTokenSymbolLength,
+				<T as pallet_pool_system::Config>::MaxTranches>,
+		>,
+	}
+	register {
+		let n in 1..<T as pallet_pool_system::Config>::MaxTranches::get();
+		let caller: <T as frame_system::Config>::AccountId = create_admin::<T>(0);
+		let tranches = build_bench_input_tranches::<T>(n);
+		let origin = RawOrigin::Signed(caller.clone());
+		prepare_asset_registry::<T>();
+		create_pool::<T>(n, caller.clone())?;
```

Comment:
I had removed this line because with it here, when running against the altair runtime, `register` below fails with a `PoolInUse` error. I'd expect this failure, since `create_pool` and `register` both end up trying to create the same pool in the underlying `pool-system` storage.

I'm still seeing that error, so I suspect this is a difference between running benches against the mocks and a real runtime?

Reply:
The `register` here actually doesn't create a pool when using the `mock` runtime. I will move `create_pool` into the `PoolMutate` mock, which should solve the problem.